File: mcp-host/src/logger.ts
/**
 * Logger module for MCP Host
 *
 * Ensures all logs go to stderr to avoid interfering with the Native Messaging protocol
 * which uses stdout for communication with the Chrome extension.
 *
 * Also provides file logging capabilities.
 */

import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';

// Log levels in order of verbosity
export enum LogLevel {
  ERROR = 0,
  WARN = 1,
  INFO = 2,
  DEBUG = 3,
}

// Get log level from environment variable or default to INFO
function getLogLevel(): LogLevel {
  const envLevel = process.env.LOG_LEVEL?.toUpperCase();

  if (!envLevel) {
    return LogLevel.DEBUG; // Default level
  }

  switch (envLevel) {
    case 'ERROR':
      return LogLevel.ERROR;
    case 'WARN':
      return LogLevel.WARN;
    case 'INFO':
      return LogLevel.INFO;
    case 'DEBUG':
      return LogLevel.DEBUG;
    default:
      return LogLevel.INFO;
  }
}

// Get log file path from environment or use default
function getLogFilePath(): string {
  const logDir = process.env.LOG_DIR || path.join(os.tmpdir(), 'mcp-host');
  const logFile = process.env.LOG_FILE || 'mcp-host.log';
  return path.join(logDir, logFile);
}

export class Logger {
  private moduleName: string;
  private static logLevel: LogLevel = getLogLevel();
  private static logFilePath: string = getLogFilePath();
  private static fileStream: fs.WriteStream | null = null;

  constructor(moduleName: string) {
    this.moduleName = moduleName;

    // Initialize file stream if not already done
    if (!Logger.fileStream) {
      try {
        // Ensure directory exists
        const dir = path.dirname(Logger.logFilePath);
        if (!fs.existsSync(dir)) {
          fs.mkdirSync(dir, { recursive: true });
        }

        // Create or open log file for appending
        Logger.fileStream = fs.createWriteStream(Logger.logFilePath, {
          flags: 'a', // Append mode
          encoding: 'utf8',
        });

        // Write header on initialization
        const startMsg = `\n[${new Date().toISOString()}] === MCP Host Logging Started ===\n`;
        Logger.fileStream.write(startMsg);

        // Set up cleanup on process exit
        process.on('exit', () => {
          if (Logger.fileStream) {
            const endMsg = `\n[${new Date().toISOString()}] === MCP Host Logging Ended ===\n`;
            Logger.fileStream.write(endMsg);
            Logger.fileStream.end();
          }
        });
      } catch (error) {
        console.error(`Failed to initialize log file at ${Logger.logFilePath}:`, error);
      }
    }
  }

  /**
   * Format a log message with timestamp, level, module name and params
   */
  private formatMessage(level: string, message: string, args: any[]): string {
    const timestamp = new Date().toISOString();
    const prefix = `[${timestamp}] [${level}] [${this.moduleName}]`;

    let formattedMessage = `${prefix} ${message}`;

    // If there are additional arguments, stringify and append them
    if (args.length > 0) {
      // Format each argument
      const formattedArgs = args
        .map(arg => {
          if (typeof arg === 'object') {
            try {
              return JSON.stringify(arg);
            } catch (e) {
              return '[Circular Object]';
            }
          }
          return String(arg);
        })
        .join(' ');

      formattedMessage += ` ${formattedArgs}`;
    }

    return formattedMessage;
  }

  /**
   * Write a log message to the file
   */
  private writeToFile(formattedMessage: string): void {
    if (Logger.fileStream) {
      try {
        Logger.fileStream.write(formattedMessage + '\n');
      } catch (error) {
        console.error(`Failed to write to log file:`, error);
      }
    }
  }

  /**
   * Log an error message (always logged)
   */
  error(message: string, ...args: any[]): void {
    if (Logger.logLevel >= LogLevel.ERROR) {
      const formattedMessage = this.formatMessage('ERROR', message, args);
      this.writeToFile(formattedMessage);
    }
  }

  /**
   * Log a warning message
   */
  warn(message: string, ...args: any[]): void {
    if (Logger.logLevel >= LogLevel.WARN) {
      const formattedMessage = this.formatMessage('WARN', message, args);
      this.writeToFile(formattedMessage);
    }
  }

  /**
   * Log an info message
   */
  info(message: string, ...args: any[]): void {
    if (Logger.logLevel >= LogLevel.INFO) {
      const formattedMessage = this.formatMessage('INFO', message, args);
      this.writeToFile(formattedMessage);
    }
  }

  /**
   * Log a debug message
   */
  debug(message: string, ...args: any[]): void {
    if (Logger.logLevel >= LogLevel.DEBUG) {
      const formattedMessage = this.formatMessage('DEBUG', message, args);
      this.writeToFile(formattedMessage);
    }
  }

  /**
   * Set log level globally for all loggers
   */
  static setLogLevel(level: LogLevel): void {
    Logger.logLevel = level;
  }

  /**
   * Get current log level
   */
  static getLogLevel(): LogLevel {
    return Logger.logLevel;
  }

  /**
   * Set log file path
   */
  static setLogFilePath(filePath: string): void {
    // Close existing stream if any
    if (Logger.fileStream) {
      Logger.fileStream.end();
      Logger.fileStream = null;
    }

    Logger.logFilePath = filePath;

    // The next logger instance created will initialize the new file stream
  }

  /**
   * Get current log file path
   */
  static getLogFilePath(): string {
    return Logger.logFilePath;
  }
}

/**
 * Create a new logger instance
 * @param moduleName The module name for this logger
 * @returns A logger instance
 */
export function createLogger(moduleName: string): Logger {
  return new Logger(moduleName);
}

File: mcp-host-go/pkg/logger/logger.go
package logger

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"sync"
	"time"
)

// LogLevel defines the severity levels for logging
type LogLevel int

const (
	// ERROR is the level for error messages
	ERROR LogLevel = iota
	// WARN is the level for warning messages
	WARN
	// INFO is the level for informational messages
	INFO
	// DEBUG is the level for debug messages
	DEBUG
)

// String returns the string representation of the log level
func (l LogLevel) String() string {
	switch l {
	case ERROR:
		return "ERROR"
	case WARN:
		return "WARN"
	case INFO:
		return "INFO"
	case DEBUG:
		return "DEBUG"
	default:
		return "UNKNOWN"
	}
}

// Logger defines the interface for logging operations
type Logger interface {
	Error(message string, args ...interface{})
	Warn(message string, args ...interface{})
	Info(message string, args ...interface{})
	Debug(message string, args ...interface{})
}

// FileLogger implements the Logger interface with file output
type FileLogger struct {
	moduleName string
	logLevel   LogLevel
	writer     io.Writer
	mutex      sync.Mutex
}

// Config holds configuration parameters for the logger
type Config struct {
	LogLevel   LogLevel
	LogDir     string
	LogFile    string
	ModuleName string
}

// NewFileLogger creates a new file logger with the given configuration
func NewFileLogger(config Config) (*FileLogger, error) {
	// Ensure log directory exists
	if err := os.MkdirAll(config.LogDir, 0755); err != nil {
		return nil, fmt.Errorf("failed to create log directory: %w", err)
	}

	// Open log file
	logPath := filepath.Join(config.LogDir, config.LogFile)
	file, err := os.OpenFile(logPath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return nil, fmt.Errorf("failed to open log file: %w", err)
	}

	// Write header
	startMsg := fmt.Sprintf("\n[%s] === MCP Host Logging Started (Module: %s) ===\n",
		time.Now().Format(time.RFC3339), config.ModuleName)
	if _, err := file.WriteString(startMsg); err != nil {
		return nil, fmt.Errorf("failed to write log header: %w", err)
	}

	return &FileLogger{
		moduleName: config.ModuleName,
		logLevel:   config.LogLevel,
		writer:     file,
	}, nil
}

// DefaultConfig returns a default logger configuration
func DefaultConfig(moduleName string) Config {
	// Get log level from environment variable
	logLevel := DEBUG
	switch os.Getenv("LOG_LEVEL") {
	case "ERROR":
		logLevel = ERROR
	case "WARN":
		logLevel = WARN
	case "INFO":
		logLevel = INFO
	case "DEBUG":
		logLevel = DEBUG
	}

	// Get log directory
	logDir := os.Getenv("LOG_DIR")
	if logDir == "" {
		logDir = filepath.Join(os.TempDir(), "mcp-host")
	}

	// Get log file name
	logFile := os.Getenv("LOG_FILE")
	if logFile == "" {
		logFile = "mcp-host.log"
	}

	return Config{
		ModuleName: moduleName,
		LogLevel:   logLevel,
		LogDir:     logDir,
		LogFile:    logFile,
	}
}

// formatMessage formats a log message with timestamp, level, module name and args
func (l *FileLogger) formatMessage(level LogLevel, message string, args ...interface{}) string {
	timestamp := time.Now().Format(time.RFC3339)
	prefix := fmt.Sprintf("[%s] [%s] [%s]", timestamp, level.String(), l.moduleName)

	formattedMessage := fmt.Sprintf("%s %s", prefix, message)

	// Format additional arguments
	if len(args) > 0 {
		formattedArgs := make([]string, len(args))
		for i, arg := range args {
			if obj, ok := arg.(interface{}); ok {
				if jsonBytes, err := json.Marshal(obj); err == nil {
					formattedArgs[i] = string(jsonBytes)
				} else {
					formattedArgs[i] = fmt.Sprintf("%+v", arg)
				}
			} else {
				formattedArgs[i] = fmt.Sprintf("%+v", arg)
			}
		}
		for _, arg := range formattedArgs {
			formattedMessage += " " + arg
		}
	}

	return formattedMessage
}

// writeToFile writes a formatted log message to the file
func (l *FileLogger) writeToFile(formattedMessage string) {
	l.mutex.Lock()
	defer l.mutex.Unlock()

	if l.writer != nil {
		_, _ = fmt.Fprintln(l.writer, formattedMessage)
	}
}

// Error logs an error message
func (l *FileLogger) Error(message string, args ...interface{}) {
	if l.logLevel >= ERROR {
		formattedMessage := l.formatMessage(ERROR, message, args...)
		l.writeToFile(formattedMessage)
	}
}

// Warn logs a warning message
func (l *FileLogger) Warn(message string, args ...interface{}) {
	if l.logLevel >= WARN {
		formattedMessage := l.formatMessage(WARN, message, args...)
		l.writeToFile(formattedMessage)
	}
}

// Info logs an informational message
func (l *FileLogger) Info(message string, args ...interface{}) {
	if l.logLevel >= INFO {
		formattedMessage := l.formatMessage(INFO, message, args...)
		l.writeToFile(formattedMessage)
	}
}

// Debug logs a debug message
func (l *FileLogger) Debug(message string, args ...interface{}) {
	if l.logLevel >= DEBUG {
		formattedMessage := l.formatMessage(DEBUG, message, args...)
		l.writeToFile(formattedMessage)
	}
}

// Close closes the log file
func (l *FileLogger) Close() error {
	l.mutex.Lock()
	defer l.mutex.Unlock()

	if closer, ok := l.writer.(io.Closer); ok {
		endMsg := fmt.Sprintf("\n[%s] === MCP Host Logging Ended (Module: %s) ===\n",
			time.Now().Format(time.RFC3339), l.moduleName)
		_, _ = fmt.Fprintln(l.writer, endMsg)
		return closer.Close()
	}
	return nil
}

// NewLogger is a convenience function that creates a logger with default config
func NewLogger(moduleName string) (Logger, error) {
	return NewFileLogger(DefaultConfig(moduleName))
}
